nomad/core_sched.go

```diff
@@ -340,6 +342,18 @@ func (c *CoreScheduler) gcEval(eval *structs.Evaluation, thresholdIndex uint64,
 	return gcEval, gcAllocIDs, nil
 }
 
+// olderVersionTerminalAllocs returns terminal allocations whose job modify index
+// is older than the job's modify index
+func olderVersionTerminalAllocs(allocs []*structs.Allocation, job *structs.Job) []string {
+	var ret []string
+	for _, alloc := range allocs {
+		if alloc.Job != nil && alloc.Job.JobModifyIndex < job.JobModifyIndex && alloc.TerminalStatus() {
```

Comment:
I think what you really want to test is the JobCreateIndex. I could change the priority of a job and that would change the job modify index. That doesn't mean we should GC its allocations. They are still potentially useful for scheduling decisions.